# Recommender Systems Workshop
*Created by Jack Douglas, UWaterloo Data Science Club*
<br>

**REMINDER: You can make a copy of this notebook by going to File > Save a copy in Drive**


## Summary

In this notebook, we will go over the implementation of a collaborative filtering recommender system. 

We will be using the MovieLens 100k dataset for this notebook. MovieLens has several datasets containing explicit ratings for thousands of movies from thousands of users. Read more about MovieLens here: https://movielens.org/. 

As well, we will be using the Surprise scikit package, which allows you to implement collaborative filtering recommender systems a few lines! Read more about Surprise here: http://surpriselib.com/.


## Packages/Libraries

We are installing the Surprise scikit package for implementing the collaborative recommender system, Pandas for creating dataframes, and NumPy for manipulating and operating on the data.

In [ ]:
!pip install scikit-surprise
import pandas as pd # Data analysis library (particularly dataframes)
import numpy as np # Computation/data analysis library
from tqdm import tqdm # Progress bar library
from surprise import Dataset 
from surprise import Reader
from surprise import SVD
from surprise.model_selection import cross_validate

     |████████████████████████████████| 11.8MB 226kB/s 
  Created wheel for scikit-surprise: filename=scikit_surprise-1.1.1-cp37-cp37m-linux_x86_64.whl size=1617632 sha256=8db8569bbb40feed1560afcc961aebd58f8b318a6d3492ce9b702658a40553ad
  Stored in directory: /root/.cache/pip/wheels/78/9c/3d/41b419c9d2aff5b6e2b4c0fc8d25c538202834058f9ed110d0
Successfully built scikit-surprise


## Dataset

As mentioned, we will be using the MovieLens 100K dataset for this notebook. It contains 100,000 ratings and 3,600 tag applications applied to 9,000 movies by 600 users.

Download the MovieLens 100K dataset here: https://files.grouplens.org/datasets/movielens/ml-latest-small.zip

Find more information on other recommender system datasets here: https://www.kdnuggets.com/2016/02/nine-datasets-investigating-recommender-systems.html

First, since we are working in Google Colab, we must mount our Google Drive to get access to all the files stored in the Google Drive.

In [ ]:
from google.colab import drive

drive.mount('/content/gdrive/')

Mounted at /content/gdrive/


We are interested in the movies and ratings database tables in the MovieLens 100k dataset. We are creating dataframes from the tables using the Pandas library.

In [ ]:
movies = pd.read_csv("/content/gdrive/MyDrive/UW/UW Data Science Club/Recommender Systems Workshop/ml-latest-small/movies.csv")
ratings = pd.read_csv("/content/gdrive/MyDrive/UW/UW Data Science Club/Recommender Systems Workshop/ml-latest-small/ratings.csv")

Here are the first five entries of the movies and ratings dataframes:




In [ ]:
movies.head()

movieId  ...                                       genres
0        1  ...  Adventure|Animation|Children|Comedy|Fantasy
1        2  ...                   Adventure|Children|Fantasy
2        3  ...                               Comedy|Romance
3        4  ...                         Comedy|Drama|Romance
4        5  ...                                       Comedy

[5 rows x 3 columns]

In [ ]:
ratings.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

Here we are revisualizing the ratings dataframe so that it looks more like the examples shown in the slides:

In [ ]:
ds = ratings.pivot(index="movieId", columns='userId', values='rating')
ds.head()

userId   1    2    3    4    5    6    7    ...  604  605  606  607  608  609  610
movieId                                     ...                                   
1        4.0  NaN  NaN  NaN  4.0  NaN  4.5  ...  3.0  4.0  2.5  4.0  2.5  3.0  5.0
2        NaN  NaN  NaN  NaN  NaN  4.0  NaN  ...  5.0  3.5  NaN  NaN  2.0  NaN  NaN
3        4.0  NaN  NaN  NaN  NaN  5.0  NaN  ...  NaN  NaN  NaN  NaN  2.0  NaN  NaN
4        NaN  NaN  NaN  NaN  NaN  3.0  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
5        NaN  NaN  NaN  NaN  NaN  5.0  NaN  ...  3.0  NaN  NaN  NaN  NaN  NaN  NaN

[5 rows x 610 columns]

## Collaborative Filtering

For the collaborative recommender system, we are using the Surprise scikit package. Here we are making a reader which has a rating scale from 1 to 5. As well, we are creating the dataset from the ratings dataframe which has only user behavioural data.

In [ ]:
reader = Reader(rating_scale=(1, 5))
# The dataframe containing the ratings. It must have three columns, corresponding to the user (raw) ids, the item(raw) ids, and the ratings, in this order.
data = Dataset.load_from_df(ratings[['userId', 'movieId', 'rating']], reader)

Recall that we are trying to have our recommender system learn the feature vectors for every movie. Surprise makes implementing this super easy! We just choose the particular algorithm we want to use to learn the feauture vectors which solve this regression problem. Here is a list of algorithms you can choose from: https://surprise.readthedocs.io/en/stable/prediction_algorithms_package.html. We have chosen the Singular Value Decomposition (SVD) algorithm, which is an algorithm which has been shown to work the best for solving recommender system regression problems.

In [ ]:
%%capture
algo = SVD() # We will use singular value decomposition (SVD) as our learning algo 
trainset = data.build_full_trainset() # Using the Surprise build_full_trainset to creating our training set
algo.fit(trainset) # Training the model (ie. learning the feature vectors)
testset = trainset.build_anti_testset() # Creating the testing set
predictions = algo.test(testset) # Making predictions on testing set

Recall that the user 1 rated movie 1 (Toy Story) highly and had no prediction for the movie 2 (Jumanji). Toy Story and Jumanji are both categorized as childrens movies with adventure and fantasy. Let's see what our prediction rating is for user 1 on Jumanji:

In [13]:
user_id = 1
movie_id = 2
for e in predictions:
  if e[0] == user_id and e[1]==movie_id:
    print(movies['title'].where(movies['movieId'] == 1)[0] + ' Prediction: ' + str(e[3]))

Toy Story (1995) Prediction: 4.0572671727279985


We will now use cross-validation to assess the accuracy of our collaborative recommender system.

In [ ]:
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True);

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    0.8714  0.8733  0.8734  0.8755  0.8735  0.8734  0.0013  
MAE (testset)     0.6715  0.6735  0.6703  0.6699  0.6706  0.6712  0.0013  
Fit time          4.91    4.88    4.88    4.85    4.89    4.88    0.02    
Test time         0.17    0.18    0.16    0.15    0.16    0.16    0.01    


If we look at the mean RMSE, that means that our predictions would likely be off by around 0.87 when used on an independent set. That means that if our recommender predicts a rating will give a movie a 3, the user will most likely be in the 2 to 4 range.